### PR TITLE
fix: gas consumption on precompile error cases

### LINKED ISSUE #413: Fix gas consumption on precompile failures
Make the gas deduction for those cases in `execute_precompile` instead of doing it inside each precompile function.

Contemplate the empty input cases (valid for some precompiles).

### PR DESCRIPTION
**Pull request by juanimedone**
Closes #413. Comes after #401.

## Summary
Changes how the consumed gas is calculated on precompile failures. Also removes `calldata`'s length check where it doesn't correspond.

## Changes
- Make new `execute_precompile` function handle the gas consumption on `PrecompileError` cases.
- Remove `calldata`'s length check in `ecrecover`, `modexp`, `ecadd` and `ecmul` and use a new auxiliary `right_pad` function to pad it with zeroes when necessary.
- Rewrite empty `calldata` unit tests.
- Add return code and gas consumption checks to all unit tests.

### CODE REVIEW

**Review by belfortep — APPROVED** [review]
LGTM! 🚀

File: src/precompiles.rs [hunk 1]
```diff
@@ -951,7 +914,6 @@ mod tests {
         let result = ecmul(&calldata, gas_limit, &mut consumed_gas);
 
         assert!(matches!(result, Err(PrecompileError::NotEnoughGas)));
-        assert_eq!(consumed_gas, gas_limit);
```
Comment (MegaRedHand) [inline src/precompiles.rs:954]: Why were these checks removed?

Comment (juanimedone) [reply src/precompiles.rs:954]: I'll make a few adjustments and put them back again.

Comment (juanimedone) [reply src/precompiles.rs:954]: Done! Now there are both return code and gas consumption checks in all tests.

**Review by MegaRedHand — APPROVED** [review]
LGTM

**Pull request merged**